docs/proposals/auth.md

```diff
@@ -0,0 +1,340 @@
+  namespace (or given sufficient information in the future, possibly their own
+  RC), while others should be allowed to push for any object in the cluster.
+
+Proposed Solution
+-----------------
+
+Heapster will adopt an authentication and authorization model based around
+querying the Kubernetes model.  Authentication will support either client
+certificate and tokens using the Kubernetes API server tokenreview
+endpoint. Authorization will be done using the Kubernetes API server
+subjectaccessreviews endpoint.
+
+### Authentication ###
+
+Authentication will support both client certificates and tokens.  Client
+certificates will be supported by checking against a configurable CA (this
+will generally be the cluster CA that gets injected into the Heapster
+pod), while tokens will be checked against the Kubernetes API server using
+a `TokenReview`.
+
+The configuration will look as such:
+
+- `--authn-ca=/path/to/ca.crt`
+- `--auth-apiserver=https://$KUBE`
+
+### Authorization ###
+
+Authorization will work by matching Heapster concepts to Kubernetes RBAC
+concepts, and then performing `SubjectAccessReview` requests against the
+API server configured in `--auth-apiserver`.
+
+The requests will use an API group of "heapster.k8s.io", and resources
+will be referred to as "resourcename.resourcegroup" (e.g.
+"pod.legacy.k8s.io")
+
+#### Historical Queries #####
+
+**Verb**: get / **Subresource**: historical-metrics
+
+Namespace and Pod requests map to their respective resources.  Requests to
+get container metrics would map a request to get metrics on the
+corresponding pod.  Requests for node and system container metrics would
+simply map to cluster permissions to get historical-metrics on nodes.
+
+Getting by Pod UID would require permissions to get historical metrics on
+pods in any namespace, since we cannot tell namespace from the pod UID.
+
+Geting pods by a list would require making multiple requests, although for
+the sake of performance, it may be desirable to just merge them into
+`namespace=$NS, resource=pod, name=""` request (and, similarly to below
+with push metrics, not fall back on individual requests if the review
+fails in order to avoid added stress on Heapster and the cluster).
+
+##### Examples #####
+
+```
+GET /api/v1/historical/namespaces/somens/pods/somepod/metrics/cpu/usage_rate
+
+apiVersion: authorization/v1beta1
+kind: SubjectAccessReview
+spec:
+    resourceAttributes:
+        verb: get
+        group: "heapster.k8s.io"
+        version: "v1"
+        namespace: somens
+        resource: "pod.legacy.k8s.io"
+        subresource: "historical-metrics"
+        name: somepod
+    user: $USERNAME
+```
+
+```
+GET /api/v1/historical/namespaces/somens/metrics/cpu/usage_rate
+
+apiVersion: authorization/v1beta1
+kind: SubjectAccessReview
+spec:
+    resourceAttributes:
+        verb: get
+        group: "heapster.k8s.io"
+        version: "v1"
+        resource: "namespace.legacy.k8s.io"
```

Comment:
to match normal authz rules, this should have the namespace and the name set to `somens`